src/SymWoodburyMatrices.jl

```diff
@@ -0,0 +1,158 @@
+  calc_inv(O.A, O.B, full(O.D));
+
+\(W::SymWoodbury, R::StridedVecOrMat) = inv(W)*R
+
+"""
+    partialInv(O)
+
+Get the factors (X,Z) in W + XZXᵀ where W + XZXᵀ = inv( A + BDBᵀ )
+"""
+function partialInv(O::SymWoodbury)
+  X = (O.A)\O.B;
+  invD = -1*inv(O.D);
+  Z = inv(invD - O.B'*X);
+  return (X,Z);
+end
+
+function liftFactorVars(A,B,D)
+  A  = sparse(A)
+  B  = sparse(B)
+  Di = sparse(inv(D))
+  n  = size(A,1)
+  k  = size(B,2)
+  M = [A    B   ;
+       B'  -Di ];
+  M = lufact(M) # ldltfact once it's available.
+  return x -> (M\[x; zeros(k,1)])[1:n,:];
+end
+
+function liftFactorVars(A,B,D::SparseMatrixCSC)
+  liftFactorVars(A,B,full(D))
+end
+
+"""
+    liftFactor(A)
+
+More stable version of inv(A).  Returns a function which computs the inverse
+on evaluation, i.e. `liftFactor(A)(x)` is the same as `inv(A)*x`.
+"""
+liftFactor(O::SymWoodbury) = liftFactorVars(O.A,O.B,O.D)
+
+function *{T}(O::SymWoodbury{T}, x::Union{Matrix,Vector,SubArray}) 
+  o = O.A*x;
+  plusBDBtx!(o, O.B, O.D, x)
+  return o
+end
+
+function plusBDBtx!(o, B, D, x)
+	o[:] = o + B*(D*(B'x))
+end
+
+# Optimization - use specialized BLAS package 
+function plusBDBtx!(o, B::Array{Float64,2}, D, x::Array{Float64,2})
+  w = D*gemm('T','N',B,x);
+  gemm!('N','N',1.,B,w,1., o)
+end
+
+Base.Ac_mul_B{T}(O1::SymWoodbury{T}, x::AbstractVector{T}) = O1*x
+Base.Ac_mul_B{T}(O1::SymWoodbury{T}, x::AbstractMatrix{T}) = O1*x
+
++(O::SymWoodbury, M::SymWoodbury)    = SymWoodbury(O.A + M.A, [O.B M.B],
+                                                   cat([1,2],O.D,M.D) );
+*(α::Real, O::SymWoodbury)           = SymWoodbury(α*O.A, O.B, α*O.D);
+*(O::SymWoodbury, α::Real)           = SymWoodbury(α*O.A, O.B, α*O.D);
++(M::AbstractMatrix, O::SymWoodbury) = SymWoodbury(O.A + M, O.B, O.D);
++(O::SymWoodbury, M::AbstractMatrix) = SymWoodbury(O.A + M, O.B, O.D);
+Base.size(M::SymWoodbury)            = size(M.A);
+Base.size(M::SymWoodbury, i)         = (i == 1 || i == 2) ? size(M)[1] : 1
+
+Base.full{T}(O::SymWoodbury{T})      = full(O.A) + O.B*O.D*O.B'
+Base.copy{T}(O::SymWoodbury{T})      = SymWoodbury(copy(O.A), copy(O.B), copy(O.D))
+
+function square(O::SymWoodbury)
+  A = O.A^2
+  D = O.D
+  B = O.B
+  Z = [(O.A*B + B) (O.A*B - B) B]
+  SymWoodbury(A, Z, full( cat([1,2],D/2,-D/2, D*B'*B*D) ) )
+end
+
+"""
+The product of two SymWoodbury matrices will generally be a Woodbury Matrix,
+except when they are the same, i.e. the user writes A'A or A*A' or A*A.
+
+Z(A + B*D*Bᵀ) = ZA + ZB*D*Bᵀ
+
+This package will not support support left multiplication by a generic 
+matrix, to keep return types consistent.
+"""
+function *(O1::SymWoodbury, O2::SymWoodbury)
+  if (O1 === O2) 
+    return square(O1)
+  else
+    if O1.A == O2.A && O1.B == O2.B && O1.D == O2.D
+      return square(O1)
+    else
+      throw(MethodError("The product of two nonidentical SymWoodbury matrices is not necessarily SymWoodbury."))
```

Comment:
How about something more useful to the user, like `"to multiply two non-identical SymWoodbury matrices, first convert to Woodbury"`? Somewhat similar to:
```jl
julia> A = sprand(5,5,0.8);
                                                                                                                                                                                                                                                                                             
julia> inv(A)
ERROR: The inverse of a sparse matrix can often be dense and can cause the computer to run out of memory. If you are sure you have enough memory, please convert your matrix to a dense matrix.
```
This both explains the problem but also how to work around the error.